File: java_version/engine/src/org/fireflow/engine/taskinstance/DynamicAssignmentHandler.java
/**
 * Copyright 2007-2008 非也
 * All rights reserved. 
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation。
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses. *
 */
package org.fireflow.engine.taskinstance;

import org.fireflow.engine.taskinstance.IAssignmentHandler;
import java.util.List;
import org.fireflow.engine.EngineException;
import org.fireflow.engine.impl.TaskInstance;
import org.fireflow.kernel.KernelException;

/**
 *
 * @author 非也,nychen2000@163.com
 */
public class DynamicAssignmentHandler implements IAssignmentHandler{
    boolean needClaim = false;
    List actorIdsList = null;
    
    public void assign(IAssignable asignable, String performerName) throws EngineException, KernelException {
        if (actorIdsList==null || actorIdsList.size()==0){
            TaskInstance taskInstance = (TaskInstance)asignable;
            throw new EngineException(taskInstance.getProcessInstanceId(),taskInstance.getWorkflowProcess(),
                    taskInstance.getTaskId(),"Actor id list can not be empty");
        }
        for (int i=0;i<actorIdsList.size();i++){
            asignable.asignToActor((String)actorIdsList.get(i), needClaim);
        }
    }

    public List getActorIdsList() {
        return actorIdsList;
    }

    public void setActorIdsList(List actorIdsList) {
        this.actorIdsList = actorIdsList;
    }

    public boolean isNeedClaim() {
        return needClaim;
    }

    public void setNeedClaim(boolean needSign) {
        this.needClaim = needSign;
    }

    
}

File: dot-net_version/FireWorkflow.Net/FireWorkflow.Net/Engine/Taskinstance/DynamicAssignmentHandler.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using FireWorkflow.Net.Model;
using FireWorkflow.Net.Engine.Impl;

namespace FireWorkflow.Net.Engine.Taskinstance
{
    /// <summary>
    /// 动态任务分配句柄，用于指定后续环节的操作员。
    /// </summary>
    public class DynamicAssignmentHandler : IAssignmentHandler
    {
        /// <summary>工作项是否需要签收</summary>
        Boolean needClaim = false;

        /// <summary>操作员Id列表</summary>
        List<String> actorIdsList = null;

        public void assign(IAssignable asignable, String performerName)// throws EngineException, KernelException 
        {
            if (actorIdsList == null || actorIdsList.Count == 0)
            {
                TaskInstance taskInstance = (TaskInstance)asignable;
                throw new EngineException(taskInstance.getProcessInstanceId(), taskInstance.getWorkflowProcess(),
                        taskInstance.getTaskId(), "actorIdsList can not be empty");
            }

            List<IWorkItem> workItems = asignable.asignToActors(actorIdsList);

            ITaskInstance taskInst = (ITaskInstance)asignable;
            if (!needClaim)
            {
                if (FormTask.ALL.Equals(taskInst.getAssignmentStrategy()) ||
                        (FormTask.ANY.Equals(taskInst.getAssignmentStrategy()) && actorIdsList.Count == 1))
                {
                    for (int i = 0; i < workItems.Count; i++)
                    {
                        IWorkItem wi = workItems[i];
                        wi.claim();
                    }
                }
            }
        }

        public List<String> getActorIdsList()
        {
            return actorIdsList;
        }

        public void setActorIdsList(List<String> actorIdsList)
        {
            this.actorIdsList = actorIdsList;
        }

        public Boolean isNeedClaim()
        {
            return needClaim;
        }

        public void setNeedClaim(Boolean needSign)
        {
            this.needClaim = needSign;
        }
    }
}
